rest/route/user.go

```diff
@@ -171,3 +172,76 @@ func (h *userAuthorGetHandler) Run(ctx context.Context) gimlet.Responder {
 
 	return gimlet.NewJSONResponse(apiAuthor)
 }
+
+type userPermissionsPostHandler struct {
+	sc          data.Connector
+	userID      string
+	permissions RequestedPermissions
+}
+
+type RequestedPermissions struct {
+	ResourceType string             `json:"resource_type"`
+	Resources    []string           `json:"resources"`
+	Permissions  gimlet.Permissions `json:"permissions"`
+}
+
+func makeModifyUserPermissions(sc data.Connector) gimlet.RouteHandler {
+	return &userPermissionsPostHandler{
+		sc: sc,
+	}
+}
+
+func (h *userPermissionsPostHandler) Factory() gimlet.RouteHandler {
+	return &userAuthorGetHandler{
+		sc: h.sc,
+	}
+}
+
+func (h *userPermissionsPostHandler) Parse(ctx context.Context, r *http.Request) error {
+	vars := gimlet.GetVars(r)
+	h.userID = vars["user_id"]
+	if h.userID == "" {
+		return errors.New("no user found")
+	}
+	permissions := RequestedPermissions{}
+	if err := util.ReadJSONInto(r.Body, &permissions); err != nil {
+		return errors.Wrap(err, "request body is not a valid Permissions request")
+	}
+	if !util.StringSliceContains(evergreen.ValidResourceTypes, permissions.ResourceType) {
+		return errors.Errorf("'%s' is not a valid resource_type", permissions.ResourceType)
+	}
+	if len(permissions.Resources) == 0 {
+		return errors.New("resources cannot be empty")
+	}
+	h.permissions = permissions
+
+	return nil
+}
+
+func (h *userPermissionsPostHandler) Run(ctx context.Context) gimlet.Responder {
+	u, err := h.sc.FindUserById(h.userID)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "can't get user for id '%s'", h.userID))
+	}
+	if u == nil {
+		return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			Message:    fmt.Sprintf("no matching user for '%s'", h.userID),
+			StatusCode: 404,
+		})
+	}
+
+	rm := evergreen.GetEnvironment().RoleManager()
```

Comment:
a bunch of the routes cache their required state/services in the handler implementations (and propogate it in the gactory,) and I kinda like that model because it makes things more explicit.